proposals/22-04-localize-command.md

```diff
@@ -0,0 +1,335 @@
+   * remote exec binaries of referenced KRM functions 
+
+   This command achieves this goal if, in the absence of remote images and custom fields in KRM
+   functions, `kustomize build` can run on the localized copy without network access.
+
+**Non-goals:**
+
+1. This command should not localize remote images or custom fields in KRM functions.
+2. This command should not copy files that the target kustomization does not reference.
+3. This command should not serve as a package manager.
+
+## Proposal
+
+The command takes the following form:
+
+<pre>
+<b>kustomize localize</b> <ins>target</ins> <ins>scope</ins> <ins>newDir</ins>
+</pre>
+
+where the arguments are:
+
+* `target`: a directory with a top-level kustomization file that kustomize will localize; can be a path to a local
+  directory or a url to a remote directory
+* `scope`: optional root directory, files outside which kustomize is not allowed to copy and localize; if not specified,
+  takes on value of `target`
+* `newDir`: destination directory of the localized copy of `target`
+
+The command creates a copy of the `target` kustomization and the local files that `target` references at `newDir`. We
+define the "files that `target` references" as:
+
+* kustomization files that `target` directly or transitively references
+* configuration files that referenced kustomization files reference
+* exec binaries of referenced KRM functions
+
+Here, configuration file means a non-kustomization yaml file. The command only copies referenced files that reside
+inside `scope`.
+
+The command localizes the copy of `target` at `newDir` by downloading all remote files that `target` references. The
+downloaded files placed in a new `localized-files` directory next to the file that referenced the downloaded files.
+Inside `localized-files`, the downloads are located on path:
+
+<pre>
+<ins>remote-host</ins> / <ins>organization</ins> / <ins>repo</ins> / <ins>version</ins> / <ins>path-to-file-in-repo</ins>
+</pre>
+
+The command rewrites remote references in `newDir` to the local paths of the downloaded files. To help ensure
+that `newDir` is a clean copy, the command additionally overwrites absolute path references into `target` to point
+to `newDir`.
+
+**Error cases**:
+* `target` does not have a top-level kustomization file
+* `scope` does not contain `target`
+* `newDir` already exists
+* `localized-files` directory already exists
+* remote reference does not have a version
+* kustomization file is malformed
+* cycle of kustomization file references exists
+
+**Warning cases**:
+* `target` references a local path that traverses outside of `scope`
+* KRM function has container image that the user might not have locally
+
+For the warning cases, the command will ignore the reference and continue execution.
+
+If the command runs without any errors or warnings, `kustomize build` without `--load-restrictor LoadRestrictionsNone`
+on `target` and `newDir` should produce the same output.
```

Comment:
+1 that's a good idea

Reply:
Will add.

Reply:
Backtracking: Will we output the result of `build` to the user, or do we check and print the diff, if there is one? The intent of this step is to either boost the user's confidence or alert them of a discrepancy if there is one, right?

Reply:
I think throwing an error with the diff is reasonable (perhaps truncating it if it is very large?)

A small caveat, if we go with the suggestion of https://github.com/kubernetes-sigs/kustomize/pull/4590#discussion_r869741199, we won't be able to do this verification step when there are KRM function exec binaries present.

Reply:
We could still verify, which would lead to `kustomize build` throwing an error from not being able to execute the KRM function exec binaries. In other words, when we run the verification, we wouldn't silence `build` errors, which would end the verification. This would also give users a better idea of what they need to do?

Reply:
> I think throwing an error with the diff is reasonable (perhaps truncating it if it is very large?)

Yes, this is what I was picturing.

Reply:
> We could still verify, which would lead to kustomize build throwing an error from not being able to execute the KRM function exec binaries. In other words, when we run the verification, we wouldn't silence build errors, which would end the verification. This would also give users a better idea of what they need to do?

I think I agree with this. That said, if we attempt the verification, I expect the initial error message will actually be about needing the alpha function enablement flags, which we absolutely definitely should not add to our verification build automatically. To make sure they know about the required permissions change(s), maybe we can emit a warning right when we make the change. 

There's at least one other case where a verification build will definitely fail: if they're using the Helm built-in (which also requires enablement flags). We might want a `--no-verify` flag because of these. Maybe verification is a beta feature?

Reply:
> We might want a --no-verify flag because of these. Maybe verification is a beta feature?

+1 to both of these suggestions